File: src/token/Operator.cpp
#include "Logika.h"
#include "Arab.h"
#include "Romawi.h"
#include "Operator.h"
#include <sstream>
#include <cassert>

Operator::Operator(){
	Opr[Kali]="*";
	Opr[Bagi]="/";
	Opr[Tambah]="+";
	Opr[Kurang]="-";
	Opr[Div]="div";
	Opr[Mod]="%";
	Opr[Kurungbuka]="(";
	Opr[Kurungtutup]=")";
	Opr[And]="&";
	Opr[Or]="|";
	Opr[Not]="~";
}

bool Operator::canConvert(const std::string& s){ 	//menembalikan true bila s dapat dikonversi ke token
	for (int i=0;i<TIPETOKEN_COUNT;i++)
		if (s==GetString((TipeToken)i))
			return true;
	return false;
}

std::string Operator::toString (const Token& T){	//mengembalikan representasi string dari token T
	assert(T.getTkn()!=Bilangan);

	return GetString(T.getTkn());
}

void Operator::SetString(TipeToken _tkn, const std::string& s){
	//I.S.:	0<= (int) _tkn <TIPETOKEN_COUNT
	//	s tidak dapat dikonversi oleh Arab, Romawi, maupun Logika
	assert(_tkn>=0 && _tkn<TIPETOKEN_COUNT);
	assert(StrRepComplies(s)); //nanti diganti jadi exception

	Opr[_tkn]=s;
}

std::string Operator::GetString(TipeToken _tkn){
	//mengembalikan Opr[_tkn]
	//prekondisi: 0<=_tkn<TIPETOKEN_COUNT
	
	assert(0<=_tkn && _tkn<TIPETOKEN_COUNT);
	return Opr[_tkn];

}

Token Operator::toToken(const std::string& s){	//mengembalikan representasi token dari string s
	assert(canConvert(s));

	Token retval;

	for (int i=0;i<TIPETOKEN_COUNT;i++)
		if (s==GetString((TipeToken)i)){
			retval.SetTkn((TipeToken)i);
			break;
		}

	return retval;
}

bool Operator::StrRepComplies(const std::string& s){
	//operator bersifat unik
	for (int i=0;i<TIPETOKEN_COUNT;i++)
		if (s==GetString((TipeToken)i))
			return false;

	//alfanumerik + simbol
	int n=s.length();
	for (int i=0;i<n;i++)
		if (	(s[i]<'0' || s[i] > '9')	&&
			(s[i]<'A' || s[i] > 'Z')	&&
			(s[i]<'a' || s[i] > 'z')	)
			switch(s[i]){
			case '+':
			case '-':
			case '/':
			case '*':
			case '%':
			case '&':
			case '|':
			case '~':
				break;
			default:
				return false;
			}
			
	//bukan merupakan representasi bilangan dalam bentuk romawi, arabic, dan logika
	Logika L; Arab A; Romawi R;
	return !(L.canConvert(s)||A.canConvert(s)||R.canConvert(s));
}

File: src/Operator.java
import java.util.EnumMap;
import java.util.HashMap;
import java.util.Set;
import java.util.Map;
import java.util.regex.Pattern;

/** kelas implementasi
* Kelas ini merupakan implementasi kelas StrToken dengan konversi sesuai aturan penulisan operator.
*/
public class Operator implements StrToken{
	private EnumMap<TipeToken, String> Opr;
	private Map<String,TipeToken> Tkn;
	
	public Operator(){
		Opr = new EnumMap<TipeToken,String>(TipeToken.class);
		Opr.put(TipeToken.Kali,"*");
		Opr.put(TipeToken.Bagi,"/");
		Opr.put(TipeToken.Tambah,"+");
		Opr.put(TipeToken.Kurang,"-");
		Opr.put(TipeToken.Div,"div");
		Opr.put(TipeToken.Mod,"mod");
		Opr.put(TipeToken.Kurungbuka,"(");
		Opr.put(TipeToken.Kurungtutup,")");
		Opr.put(TipeToken.And,"and");
		Opr.put(TipeToken.Or,"or");
		Opr.put(TipeToken.Xor,"xor");
		Opr.put(TipeToken.Not,"not");
		Opr.put(TipeToken.KurangDari,"<");
		Opr.put(TipeToken.LebihDari,">");
		Opr.put(TipeToken.KurangDariSamaDengan,"<=");
		Opr.put(TipeToken.LebihDariSamaDengan,">=");
		Opr.put(TipeToken.SamaDengan,"=");
		Opr.put(TipeToken.TidakSamaDengan,"<>");
		
		Tkn = new HashMap<String,TipeToken>();
		for (Map.Entry<TipeToken,String> entry : Opr.entrySet()){
			Tkn.put(entry.getValue(),entry.getKey());
		}		
		
	}
	
	/**mengembalikan Opr.put(_tkn]
	*/
	public String GetString(TipeToken _tkn){
		return Opr.get(_tkn);
	}

	/**mengembalikan true bila s dapat dikonversi ke token */
	public boolean canConvert(/*const*/ String s){
		return Opr.containsValue(s);
	}
	
	/**mengembalikan representasi string dari token T*/
	public String toString (/*const*/ Token T){
		return Opr.get(T.getTkn());
	}
	
	/**mengembalikan representasi token dari string s*/
	public Token toToken(/*const*/ String s){
		assert(canConvert(s));
		Token retval = new Token();
		retval.SetTkn(Tkn.get(s));
		return retval;
	}

	/**mengembalikan ekspresi reguler dari kelas turunan StrToken tersebut*/
	public String getRegEx(){
		String regex = new String();
		for (Map.Entry<String,TipeToken> entry : Tkn.entrySet()){
			regex += "(" + Pattern.quote(entry.getKey()) + ")|";
		}
		regex = regex.substring(0,regex.length()-1);
		return regex;
	}

}
